### DISCUSSION TITLE
Reading Response content w/ 400s

### DISCUSSION THREAD

**Discussion by Redoubts** [encode/httpx#2127] [Potential Issue]
I've seen some APIs that will return an error code of 400, but still put something in the response body. Usually confirming the error, or maybe a little detail about how the request was wrong.

Indeed:

```
async with httpx.AsyncClient() as client:
    google = await client.post("https://google.com")

google
<Response [405 Method Not Allowed]>

google.read()
b'...The request method <code>POST</code> is inappropriate for the URL ...'
```

But if I tack on `event_hooks={"response": [lambda x: x.raise_for_status()]}` to my client, reading this info off the response is impossible.

```
async with httpx.AsyncClient(
    event_hooks={"response": [lambda x: x.raise_for_status()]}
) as client:
    try:
        google = await client.post("https://google.com")
    except httpx.HTTPStatusError as error:
        print(error.response.read())

StreamClosed: Attempted to read or stream content, but the stream has been closed.
```

Am I doing this wrong, or should we expect `raise_for_status` to block reading?

**Comment by florimondmanca** [comment]
@Redoubts Hi,

Does accessing `error.response.text` or `error.response.content` work?

If the response has been read in the process of executing the event hook (I think the client does that) then those attributes should be available for use. Method `.read()` is only designed to be callable once in the lifecycle of a response.

**Reply by Redoubts** [reply]
Doing that fails with `ResponseNotRead: Attempted to access streaming response content, without having called "read()".` instead
